### PR TITLE
cmd/checksum: allow adding new versions to package

### PR DESCRIPTION
**Pull request by michaelkuhn**
This adds super-lazy maintainer mode to `spack checksum`: Instead of only printing the new checksums to the terminal, `-p` and `--add-to-package` will add the new checksums to the `package.py` file and open it in the editor afterwards for final checks.

### CODE REVIEW

File: lib/spack/spack/cmd/checksum.py
Comment (becker33) [inline lib/spack/spack/cmd/checksum.py]: Why not add this with `append` after we reverse the list?

**Review by tgamblin — CHANGES_REQUESTED** [review]
This looks great but it needs a test -- can you add one @michaelkuhn?

**Review by becker33 — DISMISSED** [review]
Is there any reason to print the versions in addition to writing them to the package file?

File: lib/spack/spack/cmd/checksum.py
Comment (michaelkuhn) [reply lib/spack/spack/cmd/checksum.py]: Done

**Comment by michaelkuhn** [comment]
> Is there any reason to print the versions in addition to writing them to the package file?

I just found it convenient to have the checksums printed somewhere in case I want to double-check with `git diff` afterwards. I can also remove the output if you prefer.

**Comment by michaelkuhn** [comment]
> This looks great but it needs a test -- can you add one @michaelkuhn?

To be honest, I am not really sure how to add a test for this (not very fluent in pytest and also couldn't find a test for checksum to base this on). If you have anything in mind, feel free to push to my branch. Otherwise, I can give this a try tomorrow or so.

**Comment by bernhardkaindl** [comment]
> > This looks great but it needs a test -- can you add one @michaelkuhn?
> 
> To be honest, I am not really sure how to add a test for this (not very fluent in pytest and also couldn't find a test for checksum to base this on). If you have anything in mind, feel free to push to my branch. Otherwise, I can give this a try tomorrow or so.

@tgamblin @becker33 @haampie `spack checksum --add-to-package` to add checksums to the recipe will be tested by willing human testers (us!) each day, therefore it doesn't need a regression test! ;-)

Would you force-merge it based on that?

**Comment by michaelkuhn** [comment]
I've rebased the PR to current develop, so it should apply cleanly again.

**Comment by tldahlgren** [comment]
@michaelkuhn @tgamblin Coming to the party late and possibly with a dumb question, but why add the option to `spack checksum` instead of `spack edit`, which is in a sense what you want to do?

**Comment by michaelkuhn** [comment]
> Coming to the party late and possibly with a dumb question, but why add the option to `spack checksum` instead of `spack edit`, which is in a sense what you want to do?

I just found this the most obvious place to add it, since `checksum` already contains all the logic for getting the new checksums. I have also rebased this change to current develop and tested it with a new version to make sure it works.

**Comment by tldahlgren** [comment]
@spackbot fix style

**Comment by michaelkuhn** [comment]
@tldahlgren @becker33 ping

**Review by alalazo — DISMISSED** [review]

**Comment by alalazo** [comment]
Waiting to merge until @becker33 @tgamblin confirm everything is fine with them too. In my opinion this change is good for package maintainers, and limited in impact - so anything that might go wrong / can be improved is easily addressable in a following PR.

**Comment by tgamblin** [comment]
It still needs a test.

**Comment by michaelkuhn** [comment]
Sadly, I'm still not sure how to test this. We probably don't want to modify package files while the tests run, do we? If someone has an idea or could take care of the test (since my pytest skills are lacking), that would be great.

**Comment by bernhardkaindl** [comment]
> It still needs a test.

@tgamblin Which type of test do you require?

Does the current `spack checksum` have a test?

I think @michaelkuhn should not be dealt with more test requirements than the current `spack checksum` already provides.

More than that would looks to not justified for this tiny addition.

A full test of `spack checksum` would depend on some dummy software with a test recipe (there are some recipes specifically for testing in `var/spack/recipes/...`).

For the test and as a useful feature, the code of thsd PR could be extended to have the option to write the changed recipe to a temp file and e.g. just output the unified diff from old to new to stdout, so the diff could be reviewed e.g. for batch updates and be compared against a "golden"/expected diff which the diff generated from running the test must match.

I think such variant, e.g. `spack checksum --diff` would be useful and should be easy to do, by generating the additional context for the unified diff output before and after the added line(s).

The diff output could be tested by applying it to
a copy of the test recipe, maybe with `git apply`.

In any case, I still don't see and didn't read about a justification why this needs a test.

**Comment by tgamblin** [comment]
> Does the current spack checksum have a test?

It has four.

> In any case, I still don't see and didn't read about a justification why this needs a test.

I expect this to be used frequently and honestly, we have a fair amount of bugs in Spack *already*.  The more things we put in without regression tests, the more things are going to break as development continues.  It adds up.  I don't think it's unreasonable to ask for tests, especially for changes to core (for packages we have the build farm, which we're expanding).

This is honestly the only thing that's been holding up this PR.  I like the feature, too.

**Comment by bernhardkaindl** [comment]
> > Does the current spack checksum have a test?
> 
> It has four.

Ok, then the part of getting the new version(s) is covered by test already and the new test would only have to cover the new added lines. That's good.

**Comment by michaelkuhn** [comment]
OK, I've finally managed to come up with some tests by introducing a new fixture. I hope the tests make sense. If so, this should be good to go.

**Comment by michaelkuhn** [comment]
@alalazo @tgamblin Ping? This now has tests. Anything else missing?

**Review by bernhardkaindl — APPROVED** [review]
@becker33: You requested changes to this as well, can you approve now and ping @tgamblin as well? Thanks!

**Review by alalazo — APPROVED** [review]

**Review by tgamblin — APPROVED** [review]

**Comment by tgamblin** [comment]
Thanks @michaelkuhn!

**Comment by olesenm** [comment]
Nice feature! I hope I remember the syntax.

**Pull request merged**
